File: 2023/02/main.py
import re
import math

with open('data.txt') as file:
    allowed_counts = {"red": 12, "green": 13, "blue": 14}

    answer1 = 0
    answer2 = 0
    for ln, line in enumerate(file):
        min_counts = {"red": 0, "green": 0, "blue": 0}
        for count, colour in re.findall(R"(\d+) (\w+)", line):
            min_counts[colour] = max(int(count), min_counts[colour])
        answer1 += (ln+1) * all(min_counts[c] <= allowed
                               for c, allowed in allowed_counts.items())
        answer2 += math.prod(min_counts.values())
    print(f"Part one: {answer1}")
    print(f"Part two: {answer2}")

File: 2023/02/main.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

MAX = { 'blue' => 14, 'green' => 13, 'red' => 12 }

def eval_line(line)
  MAX.keys.to_h { |colour| [colour, line.scan(/(\d+) #{colour}/).map { |n,| n.to_i }.max] }
end

part_one = 0
part_two = 0
IO.foreach('data.txt').with_index(1) do |line, game_id|
  max_cubes = eval_line(line)
  part_one += max_cubes.all? { |colour, count| count <= MAX[colour] } ? game_id : 0
  part_two += max_cubes.values.inject(:*)
end
puts "Part one: #{part_one}"
puts "Part two: #{part_two}"
